internal/store/networkpolicy.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package store
+
+import (
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kube-state-metrics/pkg/metric"
+)
+
+var (
+	descNetworkPolicyLabelsDefaultLabels = []string{"networkpolicy"}
+
+	networkpolicyMetricFamilies = []metric.FamilyGenerator{
+		{
+			Name: "kube_networkpolicy_created",
+			Type: metric.Gauge,
+			Help: "Unix creation timestamp",
+			GenerateFunc: wrapNetworkPolicyFunc(func(n *networkingv1.NetworkPolicy) *metric.Family {
+				return &metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   []string{},
+							LabelValues: []string{},
+							Value:       float64(n.CreationTimestamp.Unix()),
+						},
+					},
+				}
+			}),
+		},
+		{
+			Name: "kube_networkpolicy_labels",
+			Type: metric.Gauge,
+			Help: "Kubernetes labels converted to Prometheus labels",
+			GenerateFunc: wrapNetworkPolicyFunc(func(n *networkingv1.NetworkPolicy) *metric.Family {
+				labelKeys, labelValues := kubeLabelsToPrometheusLabels(n.Labels)
+				return &metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   labelKeys,
+							LabelValues: labelValues,
+							Value:       1,
+						},
+					},
+				}
+			}),
+		},
+		{
+			Name: "kube_networkpolicy_spec_ingress_rules",
+			Type: metric.Gauge,
+			Help: "Number of ingress rules on the networkpolicy",
+			GenerateFunc: wrapNetworkPolicyFunc(func(n *networkingv1.NetworkPolicy) *metric.Family {
+				return &metric.Family{
+					Metrics: []*metric.Metric{
+						{
+							LabelKeys:   []string{},
+							LabelValues: []string{},
+							Value:       float64(len(n.Spec.Ingress)),
```

Comment:
typically we don't do any aggregations in kube-state-metrics, why not expose a metric per rule?

Reply:
I agree this can be done afterwards very easily.

Reply:
How do you mean afterwards?

Reply:
Sorry, meant -> I agree with you @brancz as if needed the aggregation can be done later, and should not be done here, but we should expose raw metrics.

Reply:
I considered a raw metric per rule, but I was struggling to find a way to translate the `NetworkPolicy` rule spec into a sensible metric. For example, given this policy (from the docs at https://kubernetes.io/docs/concepts/services-networking/network-policies/): 
```
ingress:
  - from:
    - ipBlock:
        cidr: 172.17.0.0/16
        except:
        - 172.17.1.0/24
    - namespaceSelector:
        matchLabels:
          project: myproject
    - podSelector:
        matchLabels:
          role: frontend
    ports:
    - protocol: TCP
      port: 6379
  egress:
  - to:
    - ipBlock:
        cidr: 10.0.0.0/24
    ports:
    - protocol: TCP
      port: 5978
```

How would we sensibly represent this? The `from`/`to` blocks can have any number of sub-rules, with further nesting. I'm also not sure what the usefulness of this would be - it feels a bit like reimplementing the Kube API in metrics. Admittedly it'd be cool to be able to do things like count the number of times a particular label combination appears in NetworkPolicy rules, but I can't think how this could work 🤔 

Are there other resources we currently instrument in kube-state-metrics with complex nested structures like this?

Reply:
@tariq1890 @brancz @LiliC any thoughts on this? ☝️

Reply:
I don't think there is actually any precedence with this complexity. I think for this case I'd be ok with exposing counts, and come back to it should there be additional use cases. Can we make sure to mark those metrics as experimental?